### PR TITLE
Add finals and deprecations messages

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

Add `@final` and `@deprecations` messages.

I am targeting this branch, because this change should be done in 2.x.

### CODE REVIEW

File: tests/Block/Breadcrumb/BreadcrumbTest.php [hunk 1]
```diff
@@ -28,6 +28,9 @@
 use Twig\Environment;
 use Twig\Loader\ArrayLoader;
 
+/**
+ * @final since sonata-project/seo-bundle 2.x
+ */
 class BreadcrumbMenuBlockService_Test extends BaseBreadcrumbMenuBlockService
```
Comment (VincentLanglet) [inline tests/Block/Breadcrumb/BreadcrumbTest.php:34]: You can already add final to all the tests. We do not provide BC on them.

**Review by VincentLanglet — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
I think you need a new release with https://github.com/sonata-project/SonataBlockBundle/pull/840

**Review by core23 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
```
1x: The "Sonata\SeoBundle\Block\Social\TwitterEmbedTweetBlockService" class is considered final since sonata-project/seo-bundle 2.x. It may change without further notice as of its next major version. You should not extend it from "Sonata\SeoBundle\Tests\Fixtures\Block\TwitterEmbedTweetBSTest".
    1x in BreadcrumbTest::testBlockService from Sonata\SeoBundle\Tests\Block\Breadcrumb
```

**Review by phansys — CHANGES_REQUESTED** [review]
I think these changes are fine.
Just blocking the merge until the changelog and the upgrade notes are provided.

**Comment by wbloszyk** [comment]
> I think these changes are fine.
> Just blocking the merge until the changelog and the upgrade notes are provided.

This PR will add only messages about final, so should we mark it as `patch`?

Yes for changelog.

Upgrade note is required? We do not provide BC-break so IMO it should be done in `Upgrade-4.0`.

**Comment by phansys** [comment]
>Upgrade note is required? We do not provide BC-break so IMO it should be done in Upgrade-4.0.

From our [upgrade guide](https://github.com/sonata-project/SonataSeoBundle/blob/710022fa16105dabd7a2a264241a83d6b5f05d22/CONTRIBUTING.md#deprecation):
>Any deprecation MUST be documented in the corresponding UPGRADE-[0-9].x.md. The documentation MUST be filled inside the top unreleased section with a sub title.

**Comment by wbloszyk** [comment]
> > Upgrade note is required? We do not provide BC-break so IMO it should be done in Upgrade-4.0.
> 
> From our [upgrade guide](https://github.com/sonata-project/SonataSeoBundle/blob/710022fa16105dabd7a2a264241a83d6b5f05d22/CONTRIBUTING.md#deprecation):
> 
> > Any deprecation MUST be documented in the corresponding UPGRADE-[0-9].x.md. The documentation MUST be filled inside the top unreleased section with a sub title.



> > Upgrade note is required? We do not provide BC-break so IMO it should be done in Upgrade-4.0.
> 
> From our [upgrade guide](https://github.com/sonata-project/SonataSeoBundle/blob/710022fa16105dabd7a2a264241a83d6b5f05d22/CONTRIBUTING.md#deprecation):
> 
> > Any deprecation MUST be documented in the corresponding UPGRADE-[0-9].x.md. The documentation MUST be filled inside the top unreleased section with a sub title.

Final is not equal to deprecation. Mark code will be in next major. People can use final class but cannot extends them. Mybe we should add notes about final and iternal in upgrade guide too?

IMO In this case I should add `BaseMediaBlockService` in 3.x to allow resolve problem with extending `MediaBlockServie` in this bundle and inheritances.

File: tests/Fixtures/TwitterEmbedTweetBSTest.php [hunk 2]
```diff
@@ -15,7 +15,10 @@
 
 use Sonata\SeoBundle\Block\Social\TwitterEmbedTweetBlockService;
 
-class TwitterEmbedTweetBSTest extends TwitterEmbedTweetBlockService
+/**
+ * @deprecated since sonata-project/seo-bundle 2.x, and will be remove in 3.0.
```
Comment (phansys) [inline tests/Fixtures/TwitterEmbedTweetBSTest.php:19]: IMO, the test should not be deprecated, but marked with `@group legacy` annotation and a `NEXT_MAJOR` comment.

**Comment by phansys** [comment]
>Final is not equal to deprecation. Mark code will be in next major. People can use final class but cannot extends them. Mybe we should add notes about final and iternal in upgrade guide too?

Adding a `@final` annotation to a class doesn't make it final, it just emits a deprecation message telling that it will be declared final in a future release, if the users are extending the class.
The method responsible for this behavior is [`DebugClassLoader::checkAnnotations()`](https://github.com/symfony/error-handler/blob/035e390eb7d8c8ad58ff1649f98ebac4e79f7878/DebugClassLoader.php#L433-L437).

**Comment by VincentLanglet** [comment]
Do you have time to finish this PR @wbloszyk ?

**Review by core23 — APPROVED** [review]

**Comment by wbloszyk** [comment]
@VincentLanglet I should find some time tomorrow, or at least in this week.

**Review by VincentLanglet — COMMENTED** [review]
Shouldn't some code of the last commit be in another PR ?

The PR is talking about Adding final/deprecation messages.
But the last commit is
- requiring nyholm/psr7
- adding some `@final` annotations
- deprecating some constant
- Adding a `tweetGetter` dependency to service
- Changing a || to &&
- ...

File: composer.json [hunk 3]
```diff
@@ -21,6 +21,7 @@
     ],
     "require": {
         "php": "^7.3 || ^8.0",
+        "nyholm/psr7": "^1.2",
```
Comment (core23) [inline composer.json:24]: Not related to this PR and not necessary at all

**Review by core23 — CHANGES_REQUESTED** [review]

**Comment by wbloszyk** [comment]
> Shouldn't some code of the last commit be in another PR ?
> 
> The PR is talking about Adding final/deprecation messages.
> But the last commit is
> 
> * requiring nyholm/psr7
> * adding some `@final` annotations
> * deprecating some constant
> * Adding a `tweetGetter` dependency to service
> * Changing a || to &&
> * ...

Agree, last commit should be add as seperated PR. I have some problem with this PR becouse:
- We haven't test for `tweet getter` part
- `\GuzzleHttp\Client()` is wrong configured (should be by `__construct()` argument, not by `setConfig()`,
- `tweet` setting help says: `Tweet id, URL or content. If you put the HTML content directly, other options won\'t be handled.` But `NULL` is returned when url/id pattern is not detected.

I change `||` to `&&` becouse it i move checking pattern to other place where chaning logic is required.

I didn't check it well but IMO bug was introduce in: https://github.com/sonata-project/SonataSeoBundle/commit/5e6ca8ff6942e1c4ff84c8072309514b5628f785

**Pull request closed without merging**